sdk/src/azure/core/az_log.c

```diff
@@ -16,75 +16,51 @@
 
 #ifndef AZ_NO_LOGGING
 
-static az_log_classification const* volatile _az_log_classifications = NULL;
+// Only using volatile here, not for thread safety, but so that the compiler does not optimize what
+// it falsely thinks are stale reads.
 static az_log_message_fn volatile _az_log_message_callback = NULL;
+static az_log_should_write_fn volatile _az_log_should_write_callback = NULL;
 
-void az_log_set_classifications(az_log_classification const classifications[])
-{
-  _az_log_classifications = classifications;
-}
-
-void az_log_set_callback(az_log_message_fn az_log_message_callback)
+void az_log_set_message_callback(az_log_message_fn az_log_message_callback)
 {
   _az_log_message_callback = az_log_message_callback;
 }
 
-// _az_LOG_WRITE_engine is a function private to this .c file; it contains the code to handle
-// _az_LOG_SHOULD_WRITE & _az_LOG_WRITE.
-//
-// If log_it is false, then the function returns true or false indicating whether the message
-// should be logged (without actually logging it).
-//
-// If log_it is true, then the function logs the message (if it should) and returns true or
-// false indicating whether it was logged.
-static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
+void az_log_set_filter_callback(az_log_should_write_fn az_log_should_write_callback)
 {
-  // Copy the volatile fields to local variables so that they don't change within this function
-  az_log_message_fn const callback = _az_log_message_callback;
-  az_log_classification const* classifications = _az_log_classifications;
-
-  if (callback == NULL)
-  {
-    // If no one is listening, don't attempt to log.
-    return false;
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
-
-  for (az_log_classification const* cls = classifications; *cls != AZ_LOG_END_OF_LIST; ++cls)
-  {
-    // If this message's classification is in the customer-provided list, we should log it.
-    if (*cls == classification)
-    {
-      if (log_it)
-      {
-        callback(classification, message);
-      }
-
-      return true;
-    }
-  }
-
-  // This message's classification is not in the customer-provided list; we should not log it.
-  return false;
+  _az_log_should_write_callback = az_log_should_write_callback;
```

Comment:
@antkmsft : Didn't we have a compare exchange in the PAL? We should call it here to guarantee atomicity.

Reply:
Ah, if only we could! Logging is used by IoT, and they don't want PAL dependency, because their customers will have to provide az_noplatform.lib to linker at all times.

Reply:
Well, we need something to do this assignment atomically; volatile by itself doesn't guarantee that.

Reply:
Don't we "assume it is atomic"? You can do volatile and be reasonably good, and would cause no AVs in many environments, in this specific case, especially given that some compilers/platforms may have flags/guarantees to do that atomically.